cmd/clusterctl/pkg/client/init.go

```diff
@@ -0,0 +1,113 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"github.com/pkg/errors"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/cluster"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+)
+
+// Init a management cluster by adding the requested list of providers.
+func (c *clusterctlClient) Init(options InitOptions) ([]Components, bool, error) {
+	// gets access to the management cluster
+	cluster, err := c.clusterClientFactory(options.Kubeconfig)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// ensure the custom resource definitions required by clusterctl are in place
+	if err := cluster.ProviderInventory().EnsureCustomResourceDefinitions(); err != nil {
+		return nil, false, err
+	}
+
+	// checks if the cluster already contains a Core provider.
+	// if not we consider this the first time init is executed, and thus we enforce the installation of a core provider (if not already explicitly requested by the user)
+	firstRun := false
+	currentCoreProvider, err := cluster.ProviderInventory().GetDefaultProviderName(clusterctlv1.CoreProviderType)
+	if err != nil {
+		return nil, false, err
+	}
+	if currentCoreProvider == "" && options.CoreProvider == "" {
+		firstRun = true
+		options.CoreProvider = config.ClusterAPIName
+	}
+
+	// create and installer service, add the requested providers to the install queue (thus performing validation of the target state of the management cluster
+	// before starting the installation), and then perform the installation.
+	installer := cluster.ProviderInstaller()
+
+	addOptions := addToInstallerOptions{
+		installer:         installer,
+		targetNameSpace:   options.TargetNameSpace,
+		watchingNamespace: options.WatchingNamespace,
+		force:             options.Force,
+	}
+
+	if options.CoreProvider != "" {
+		if err := c.addToInstaller(addOptions, clusterctlv1.CoreProviderType, options.CoreProvider); err != nil {
+			return nil, false, err
+		}
+	}
+
+	if err := c.addToInstaller(addOptions, clusterctlv1.BootstrapProviderType, options.BootstrapProviders...); err != nil {
+		return nil, false, err
+	}
+
+	if err := c.addToInstaller(addOptions, clusterctlv1.InfrastructureProviderType, options.InfrastructureProviders...); err != nil {
+		return nil, false, err
+	}
+
+	r, err := installer.Install()
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Components is an alias for repository.Components; this makes conversion
+	rr := make([]Components, len(r))
```

Comment:
I think I'd vote for more descriptive variable names than `r` and `rr`

Reply:
fixed